substrate/frame/sassafras/src/lib.rs

```diff
@@ -0,0 +1,1079 @@
+				warn!(target: LOG_TARGET, "Tickets shall be proposed in the first epoch half",);
+				return InvalidTransaction::Stale.into()
+			}
+
+			// This should be set such that it is discarded after the first epoch half
+			let tickets_longevity = epoch_length / 2 - current_slot_idx;
+			let tickets_tag = tickets.using_encoded(|bytes| hashing::blake2_256(bytes));
+
+			ValidTransaction::with_tag_prefix("Sassafras")
+				.priority(TransactionPriority::max_value())
+				.longevity(tickets_longevity as u64)
+				.and_provides(tickets_tag)
+				.propagate(true)
+				.build()
+		}
+	}
+}
+
+// Inherent methods
+impl<T: Config> Pallet<T> {
+	/// Determine whether an epoch change should take place at this block.
+	///
+	/// Assumes that initialization has already taken place.
+	pub(crate) fn should_end_epoch(block_num: BlockNumberFor<T>) -> bool {
+		// The epoch has technically ended during the passage of time between this block and the
+		// last, but we have to "end" the epoch now, since there is no earlier possible block we
+		// could have done it.
+		//
+		// The exception is for block 1: the genesis has slot 0, so we treat epoch 0 as having
+		// started at the slot of block 1. We want to use the same randomness and validator set as
+		// signalled in the genesis, so we don't rotate the epoch.
+		block_num > One::one() && Self::current_slot_index() >= T::EpochLength::get()
+	}
+
+	/// Current slot index relative to the current epoch.
+	fn current_slot_index() -> u32 {
+		Self::slot_index(CurrentSlot::<T>::get())
+	}
+
+	/// Slot index relative to the current epoch.
+	fn slot_index(slot: Slot) -> u32 {
+		slot.checked_sub(*Self::current_epoch_start())
+			.and_then(|v| v.try_into().ok())
+			.unwrap_or(u32::MAX)
+	}
+
+	/// Finds the start slot of the current epoch.
+	///
+	/// Only guaranteed to give correct results after `initialize` of the first
+	/// block in the chain (as its result is based off of `GenesisSlot`).
+	fn current_epoch_start() -> Slot {
+		Self::epoch_start(EpochIndex::<T>::get())
+	}
+
+	/// Get the epoch's first slot.
+	fn epoch_start(epoch_index: u64) -> Slot {
+		const PROOF: &str = "slot number is u64; it should relate in some way to wall clock time; \
+							 if u64 is not enough we should crash for safety; qed.";
+
+		let epoch_start = epoch_index.checked_mul(T::EpochLength::get() as u64).expect(PROOF);
+		GenesisSlot::<T>::get().checked_add(epoch_start).expect(PROOF).into()
+	}
+
+	pub(crate) fn update_ring_verifier(authorities: &[AuthorityId]) {
+		debug!(target: LOG_TARGET, "Loading ring context");
+		let Some(ring_ctx) = RingContext::<T>::get() else {
+			debug!(target: LOG_TARGET, "Ring context not initialized");
+			return
+		};
+
+		let pks: Vec<_> = authorities.iter().map(|auth| *auth.as_ref()).collect();
+
+		debug!(target: LOG_TARGET, "Building ring verifier (ring size: {})", pks.len());
+		let verifier_data = ring_ctx
+			.verifier_data(&pks)
+			.expect("Failed to build ring verifier. This is a bug");
```

Comment:
Can we have some kind of default value for this to not panic?

Reply:
There is no proper `Default` for `RingVerifierData`.  But, we may think about log an `error!` and store a `None` in the `RingVerifierData` storage.

At this point the protocol will fallback to the AURA like mechanism... allowing a recovery once that we are able to construct the verifier.

WDYT?